Add VP9 dual pipe encoder support for KBL

sreerenjb opened the pull request:
The patch series is adding VP9 dual pipe encoder support for GEN9. It uses the existing media-kernel for gen9. Some of the parameter enable/disable is based on the legacy i965 driver implementation. Dynamic resolution change is not supported in this patch series.

sreerenjb committed:
Encoder_VP9: Add infrastructure for vme+pak vp9 enc enablement

-- Add the CODECHAL_VP9 to the CscDS kernel functions
-- Define CODEC_VP9_MAX_SEGMENTS as a common param for dec and enc
-- Setup encode interface functions
-- Add VP9 config params in linux common and ddi layer
-- Conditional compilation enabled

sreerenjb committed:
Encoder_VP9: Extend the hw commands for gen9

This patch extending the below hw commands:
HCP_VP9_PIC_STATE_CMD
  Add DW19 to DW33

HCP_VP9_SEGMENT_STATE_CMD
  Add DW7

HCP_IND_OBJ_BASE_ADDR_STATE_CMD
  Add DW14 to DW27

sreerenjb committed:
Add VP9 Encoder for Gen9/KBL

Add the dual pipe (VME+PAK) VP9 encoder support for KBL platform.
The current media-kernel which we ship for gen9 platforms already
has the vp9 ENC/VME in it.This patch is utilizing the existing
media-kernels for HME, ME and MBEnc and a fixed function pak for
block transformation, quantization and final bit packing.

rosetta-jpn commented:
Thanks for this feature. When is this feature expected to be checked in?

jbarnes993 commented:
There's a Chromium bug for this too, but fwiw it would be either nice to have this or have the Intel media team support the i965 driver on KBL.  I think we'd prefer this CL though, since it seems to be fairly efficient relative to i965, at least for VP9.  Talk to hiroh for more details.

rosetta-jpn commented:
Sree, does this patch work for CML and WHL too, not only KBL?

sreerenjb commented:
Hiro, I haven't tested on CML & WHL. Ideally, similar code base should work for all since they share the same exact GEN9 graphics. But, since I limited some of the HW command extension in the patch only to kbl (eg: https://github.com/intel/media-driver/pull/717/commits/c00c1ec88de10c4e3e47f790f7dfe719ccfe6337), it many not work as it is on CML. We may move the kbl specific codebase to base classes in gen9 directory.

sreerenjb committed:
VP9_ENC/GEN9: Relax the compressed frames minimum size restriction.

The current implementation of the CQP mode encode has a
compressed buffer size restriction of 4Kb. The restriction
doesn't make sense for low-resolution & low-bitrate
video encoding use cases since inter-frames can be encoded
with much less number of bytes.

For eg: most of the P-frames in a video-conferencing session
of 640x480 resolution at 750kbps require only 1Kb bitrate
allocation and currently the hardware PAK unit fill the rest
of the 3Kb (4Kb-1Kb) with zeros. This will jeopardize the
quality & bit-rate allocation of the hybrid CBR encode
models[1] where we run the bitrate control algorithm in
software clubbed with a CQP hardware encoder:
[1]: https://cgit.freedesktop.org/~sree/gstreamer-vaapi/commit/?h=VP9_SVC_and_SoftwareBRC&id=355f16bdb32965425b274aa082c1334b0edd5992).

sreerenjb committed:
VP9_ENC/GEN9: Fix odd dimension video encode

The PIC_STATE hardware command buffer was getting
filled with wrong width/height values for odd resolutions
which caused gpu hang. This patch will enable support for
odd resolution video encode.

sreerenjb committed:
media_libva_caps_g10.cpp: Fix redeclaraion of AttribMap

karoldob commented:
I've tested this patch on Ubuntu 21.04, merged into media-driver 21.1.1 branch,
Results:
on Intel gen10 - all works fine
on Intel gen11 - encoding process has no erros, but wideo is completely artifacted.

sreerenjb commented:
Gen11 has a fixed function VP9 encode and the iHD driver officially supports it. This patch is not intended for the full fixed function solution.

IK4MS commented:
Can confirm it appears to work fine on CML (10700K), at least with the few settings that I used.

ccorbacho commented:
Is there any chance of this being reviewed and merged? As it's been open for 6 years now, but is still useful for those of us with older hardware. @XinfengZhang ?
